{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Machine learning application in cancer prognosis\n",
    "\n",
    "This project is far from finished but I would like to use it as an opportunity to demonstrate my skills in python and machine learning.\n",
    "\n",
    "**Objective**: Create a machine learning algorith to predict cancer prognosis for patients with colorectal adenocarcinoma.\n",
    "\n",
    "**Data**: genomic and clinical data of Colon adenocarcinoma (COAD) cohort obtained from The Cancer Genome Atlas (TCGA). The results shown here are in whole based upon data generated by TCGA Research Network: https://www.cancer.gov/tcga.\n",
    "\n",
    "\n",
    "\n",
    "Loading of the libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import csv\n",
    "from numpy.random import randn\n",
    "import matplotlib as plt\n",
    "import seaborn as sns\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/gv1u14/Desktop\n",
      "/Users/gv1u14/Desktop/ML\n"
     ]
    }
   ],
   "source": [
    "path = os.getcwd()\n",
    "print(path)\n",
    "os.chdir('/Users/gv1u14/Desktop/ML')\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading the data that was proviously pre-processed for different project. Code is not shown here becuase the analysis was done in R.\n",
    "\n",
    "datE contains genomic data\n",
    "\n",
    "datS contains clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "datE = pd.read_csv(\"datE.csv\")\n",
    "datS = pd.read_csv(\"datS_COAD_341.csv\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for col in datS.columns:\n",
    "#    print(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(341, 133)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datS.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparing the data for the analysis\n",
    "\n",
    "TCGA provides a lot of clinical data for each patient, however, missing values and categorical values are present."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Numeric variables\n",
    "Missing vaues in **numerical continous variable** (i.e. 'mitoScore','age','bmi','stemness_DNAmeth_based','Tumour cell DNA fraction', 'Fraction of genome with subclonal SCNAs', MS_stage2 subcategorties,) were replaced by **mean values** of each variable.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "list1 = ['mitoScore','age','bmi','stemness_DNAmeth_based','abs_purity','abs_ploidy','WGD','Tumor cell DNA fraction','Fraction of genome with duplicated alleles','CDS','Fraction of genome with subclonal SCNAs','N_homozygous_deletions']                                                \n",
    "list2 = [col for col in datS.columns if 'MS_' in col]\n",
    "list3 = list1 + list2\n",
    "for x in list3:\n",
    "    datS[[x]] = datS[[x]].fillna(value=datS[[x]].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Missing vaues in **numerical discrete variable** (i.e. cens, CDKN2A methylation,MHL1 methylation) were replaced by **median value**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "list4 = ['cens','CDKN2A methylation','MLH1 methylation']\n",
    "for x in list4:\n",
    "    datS[[x]] = datS[[x]].fillna(value=datS[[x]].median())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Removing variables\n",
    "\n",
    "In this analysis progression-free interval (i.e. PFI_yrs) will be used as a dependent variabel and therefore patients with unknown PFI time are removed.\n",
    "\n",
    "From datS were removed columns 'project','weight','height','bmi_bin','country','geographic region','organ','gastric history','anatomic region','time_yrs', 'Days to last known alive','Left or right colon','vital statuts' due to >90% proportion of NA, insignificant value of the variable or duplicity of the information.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "datS = datS.dropna(subset = ['PFI_yrs'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "col = ['project','weight','height','bmi_bin','Country','Geographic Region','Organ','Gastric histological classification','Anatomic Region','time_yrs', 'Days to last known alive','Left or right colon','Vital status.1','Pathologic Stage']\n",
    "datS = datS.drop(col, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Categorical variables\n",
    "After all numerical NAs were handled, missing values of categorical variables were replaced by the most frequent value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "datS = datS.fillna(datS.mode().iloc[0])\n",
    "datS.isnull().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handling ordinal categorical variables\n",
    "\n",
    "**Oridinal** categorical variables (e.g.can be ordered): *Pathologic T,Pathologic N, Pathoogogic M, Stage*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['T2' 'T4a' 'T3' 'T4b' 'T4' 'T1' 'Tis']\n"
     ]
    }
   ],
   "source": [
    "print(datS['Pathologic T'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "becuase pathological T4 can be divided into T4a and T4b, I will replace these two subclasses by T4. This might lead to loss of information but it makes biologicaly more sense than having three T4 classes. Same process will be applied to pathological N and M variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "datS['Pathologic T'] = datS['Pathologic T'].replace(to_replace=['T4a','T4b'],value='T4')\n",
    "\n",
    "T_map = {'Tis':0,'T1':1,'T2':2,'T3':3,'T4':4}\n",
    "datS['Pathologic T'] = datS['Pathologic T'].map(T_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['N0' 'N1b' 'N2b' 'N2' 'N1' 'N2a' 'N1a' 'N1c' 'NX']\n"
     ]
    }
   ],
   "source": [
    "print(datS['Pathologic N'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "datS['Pathologic N'] = datS['Pathologic N'].replace(to_replace=['N1a','N1b','N1c'],value='N1')\n",
    "datS['Pathologic N'] = datS['Pathologic N'].replace(to_replace=['N2a','N2b'],value='N2')\n",
    "\n",
    "N_map = {'NX':0,'N0':1,'N1':2,'N2':3}\n",
    "datS['Pathologic N'] = datS['Pathologic N'].map(N_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['M0' 'MX' 'M1b' 'M1a' 'M1']\n"
     ]
    }
   ],
   "source": [
    "print(datS['Pathologic M'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "datS['Pathologic M'] = datS['Pathologic M'].replace(to_replace=['M1a','M1b'],value='M1')\n",
    "\n",
    "M_map = {'MX':0,'M0':1,'M1':2}\n",
    "datS['Pathologic M'] = datS['Pathologic M'].map(M_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['I' 'III' 'II' 'IV']\n"
     ]
    }
   ],
   "source": [
    "print(datS['Stage'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "Stage_map = {'I':0,'II':1,'III':2,'IV':3}\n",
    "datS['Stage'] = datS['Stage'].map(Stage_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Handling nominal categorical variables\n",
    "\n",
    "**nominal** categorical variables: *'Race','Gender','Hypermethylation category','anatomic_site',colorectal_CMS',Molecular subtype,MSI status, Vital status*\n",
    "\n",
    "One-hot encoding will be used to handle categorical variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "nom_cat = ['Race','Gender','MSI Status','Molecular_Subtype','Vital status','colorectal_CMS','Hypermethylation category','anatomic_site','Vital status']\n",
    "one_hot = pd.get_dummies(datS[nom_cat],drop_first=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "datS = datS.drop(nom_cat,axis = 1)\n",
    "datS = datS.join(one_hot)\n",
    "datS.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
